saleor/graphql/attribute/mutations.py

```diff
@@ -747,14 +748,16 @@ def success_response(cls, instance):
 
     @classmethod
     def post_save_action(cls, info, instance, cleaned_input):
-        variants = product_models.ProductVariant.objects.filter(
-            Exists(instance.variantassignments.filter(variant_id=OuterRef("id")))
-        )
-
-        product_models.Product.objects.filter(
-            Q(Exists(instance.productassignments.filter(product_id=OuterRef("id"))))
-            | Q(Exists(variants.filter(product_id=OuterRef("id"))))
-        ).update(search_index_dirty=True)
+        with transaction.atomic():
+            variants = product_models.ProductVariant.objects.filter(
+                Exists(instance.variantassignments.filter(variant_id=OuterRef("id")))
+            )
+            qs = product_models.Product.objects.select_for_update(of=("self",)).filter(
+                Q(Exists(instance.productassignments.filter(product_id=OuterRef("id"))))
+                | Q(Exists(variants.filter(product_id=OuterRef("id"))))
+            )
+            list(qs)  # evaluate qs to enforce acquiring the lock.
```

Comment:
reworked this to use a subquery so that we don't fetch the rows at all since it's not needed (we just need proper locking to take place upfront of actual update)